File: internal/php-lint-run.rb
require 'parallel'
require 'open3'

# original)
#fd -HI --type f -E .git -E vendor '.*\.php' | sed -u -E 's/\\/\//g' | \
#  xargs -I'{}' -n1 php -l {} 2>&1 1>/dev/null | \
#  grep --line-buffered -v -E '^No syntax errors'


def main
  cmd = 'fd -HI --type f -E .git -E vendor ".*\.php$" | sed -E "s/\\\\\\\\/\\\\//g"'
  files = `#{cmd}`.lines.map {|file| file.chomp}

  threads = Parallel.processor_count
  Parallel.each_with_index(files, in_threads: threads) do |file, i|
    Open3.popen3("php -l #{file}") do |_, o, e|
      # check have error
      error = true
      o.each do |line|
        if line.index('No syntax errors') != nil
          error = false
          break
        end
      end

      msg = e.first
      stablePrint(i, error, msg)
    end
  end
end


$msgs = {}
$cur_i = 0

# 順番を保ったまま出力
def stablePrint(i, error, msg)
  # メッセージ更新
  $msgs[i] = {
    error: error,
    msg: msg
  }

  # メッセージ表示
  while $msgs.has_key?($cur_i)
    msg_ = $msgs[$cur_i]
    if msg_[:error]
      print msg_[:msg]
    end

    $msgs.delete($cur_i)
    $cur_i += 1
  end
end

main

File: internal/php-lint-run.py
# -*- coding: utf-8 -*-

# original)
#fd -HI --type f -E .git -E vendor '.*\.php' | sed -u -E 's/\\/\//g' | \
#  xargs -I'{}' -n1 php -l {} 2>&1 1>/dev/null | \
#  grep --line-buffered -v -E '^No syntax errors'

import os
import re
import subprocess

from concurrent.futures import ThreadPoolExecutor
from threading import Lock



def main():
    cmd = 'fd -HI --type f -E .git -E vendor ".*\.php$" | sed -E "s/\\\\\\\\/\\\\//g"'
    files, _ = run(cmd)

    with ThreadPoolExecutor(max_workers=os.cpu_count()) as executer:
        executer.map(doLint, zip(range(len(files)), files))


lock = Lock()

# linter処理
def doLint(tpl):
    i, file = tpl

    o, e = run('php -l {}'.format(file))

    error = not any(map(lambda line: re.search('^No syntax errors', line), o))
    msg = e[0] if len(e) > 0 else ''

    global lock
    with lock:
        stablePrint(i, error, msg)

# 外部コマンドを実行し、出力を戻り値で返す
def run(cmd):
    outs = []
    errs = []

    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    while True:
        out = proc.stdout.readline()
        err = proc.stderr.readline()
        if not out and not err:
            break

        outs.append(out.decode('utf-8').rstrip())
        errs.append(err.decode('utf-8').rstrip())

    return outs, errs


msgs = {}
cur_i = 0
# 順番を保ったまま出力
def stablePrint(i, error, msg):
    global msgs
    global cur_i

    # メッセージ更新
    msgs[i] = {
        'error': error,
        'msg': msg
    }

    # メッセージ表示
    while cur_i in msgs:
        msg = msgs[cur_i]
        if msg['error']:
            print(msg['msg'])

        del msgs[cur_i]
        cur_i += 1


if __name__ == '__main__':
    main()
